### PR TITLE
2 4 user creation

### PR DESCRIPTION
**Pull request by naphets123**

### CODE REVIEW

File: library/ajax/testdbdelete.php [hunk 1]
```diff
@@ -13,9 +15,9 @@
         $ids = [];
         // get ids of user by emails
         if (isset($_POST['emails'])) {
-            $emails = explode(',', $_POST['emails']);
+            $emails = $_POST['emails'];
```
Comment (HaGuesto) [inline library/ajax/testdbdelete.php]: Did you check if other tests are using this function? You are just passing single addresses to the `testdbdelete` function, but these lines of code were meant to also work if multiple addresses are passed to `$_POST['emails']`. Did you check that?

File: library/ajax/testdbdelete.php [hunk 2]
```diff
@@ -41,11 +43,12 @@
             LEFT JOIN cms_usergroups g ON u.cms_usergroups_id = g.id
             WHERE g.organisation_id = 100000000'
         );
-
+        $allowed = array_column($allowed[$_POST['table']], 'id');
```
Comment (HaGuesto) [inline library/ajax/testdbdelete.php]: What is this for?

File: library/ajax/testdbdelete.php [hunk 3]
```diff
@@ -41,11 +43,12 @@
...
         // Test if table is key in $allowed and the ids can be deleted
         foreach ($ids as $id) {
-            if ($return['success'] && isset($allowed[$_POST['table']]) && in_array($id, $allowed[$_POST['table']])) {
+            if ($return['success'] && isset($allowed[$_POST['table']]) && in_array($id, $allowed)) {
```
Comment (HaGuesto) [inline library/ajax/testdbdelete.php]: Why did you change this? `$_POST['table']` should be equal to 'cms_users'. All data which is allowed to be deleted is specified in line 40.

By your change you are hacking a security check.

**Comment by HaGuesto** [comment]
I just checked the testplan and left comments @naphets123.

File: cypress/integration/2_4_CreateUser.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        cy.get("button").contains("Save and close").click()
+        cy.get("tr").contains(Testaddress).should("be.visible")
+    })
+
+    it("2_4_1 Create New User without axpiry date", () => {
+        DeleteTestUser(Testaddress)
+        FillForm(Testname, Testaddress, Testgroup)
+        cy.get("button").contains("Save and close").click()
+        cy.get("tr").contains(Testaddress).should("be.visible")
+
```
Comment (HaGuesto) [inline cypress/integration/2_4_CreateUser.js]: empty line

File: cypress/integration/2_4_CreateUser.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    it("2_4_3 Create New User with current date", () => {
+        DeleteTestUser(Testaddress)
+        FillForm(Testname,Testaddress,Testgroup)
+        let today = new Date()
+        let current_date = today.getDate() + "-" + today.getMonth() + "-"+today.getFullYear();
+        cy.get("input[data-testid='user_valid_to']").type(current_date);
+        cy.get("button").contains("Save and close").click()
+        cy.get("ul[class='pagemenu list-unstyled']").children('li').last().click()
+        cy.get("a").contains("Deactivated").click()
+
```
Comment (HaGuesto) [inline cypress/integration/2_4_CreateUser.js]: empty line

**Comment by HaGuesto** [comment]
btw 5 out 6 tests fail locally for me even after a fresh reseed.

image
image

File: library/ajax/testdbdelete.php (hunk 2 above)
Comment (naphets123) [reply library/ajax/testdbdelete.php]: The current structure with db_value returns an array of arrays, with the inner array containing the information we are interested in. Therefore i use array_column to get a single array with the desired information in it

Comment (HaGuesto) [reply library/ajax/testdbdelete.php]: I see what you mean. My bad. It seems I did not check exactly how the db ids are collected. 

However, instead of this hack could you rather replace `db_value` and `db_array` with `db_query` and `db_fetch` where the data is collected?

Comment (HaGuesto) [reply library/ajax/testdbdelete.php]: @naphets123 check this to understand how db_fetch can be used. https://github.com/boxwise/dropapp/blob/85ec2e2647d061e63b13052b654a6613bec5dc2a/library/lib/database.php#L63

Comment (naphets123) [reply library/ajax/testdbdelete.php]: ok i think you could reuse the db_fetch in this way but it feels like then i am basically rewriting an alternative db_array anyway?

File: library/ajax/testdbdelete.php (hunk 1 above)
Comment (naphets123) [reply library/ajax/testdbdelete.php]: Is the issue here really single vs multiple? i thought it's more passing an array vs a string

**Comment by naphets123** [comment]
The tests fail because the last assertion is borken, even when the delete is successful as i told before. Commenting it out in cypress/support/database will have them run successfully.

**Comment by HaGuesto** [comment]
[Cypress: Automated Testing of User Management:2.4 User Creation](https://trello.com/c/9YKVydX0/314-cypress-automated-testing-of-user-management24-user-creation)

File: cypress/integration/2_4_CreateUser.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    beforeEach(function () {
+        cy.loginAsAdmin();
+        cy.visit('/?action=cms_users_edit&origin=cms_users');
+
+    })
+    
+    it("2_4_0_A Check empty form + empty submit",() => {
+        cy.get("input[data-testid='user_name']").should("be.visible")
+        cy.get("input[data-testid='user_email']").should("be.visible")
+        cy.get("span[id='select2-chosen-1']").should("be.visible")
```
[the remark this answers is not in the document]
Comment (HaGuesto) [reply cypress/integration/2_4_CreateUser.js]: @naphets123 I still do not have an answer for this.

File: library/ajax/testdbdelete.php [hunk 7]
```diff
@@ -13,39 +15,41 @@
...
+        ), 'id');
+        $allowed['cms_users'] = array_column(db_array(
             'SELECT u.id
             FROM cms_users u
             LEFT JOIN cms_usergroups g ON u.cms_usergroups_id = g.id
             WHERE g.organisation_id = 100000000'
-        );
-
+        ), 'id');
+        //$allowed = array_column($allowed[$_POST['table']], 'id');
```
Comment (HaGuesto) [inline library/ajax/testdbdelete.php]: @naphets123 remove this line.

**Comment by HaGuesto** [comment]
closed due to #311
